File: src/acceptance-test/java/net/slipp/qna/QuestionPage.java
package net.slipp.qna;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.collect.Lists;

public class QuestionPage {
	private static final Logger log = LoggerFactory.getLogger(QuestionPage.class);
	
    private WebDriver driver;

    public QuestionPage(WebDriver driver) {
        this.driver = driver;
    }

    public QuestionPage(WebDriver driver, String title) {
        this.driver = driver;
        assertThat(driver.getTitle(), is(title));
    }

    public void verify(String title, String contents, String plainTags) {
        WebElement titleElement = driver.findElement(By.cssSelector(".qna-title"));
        assertThat(titleElement.getText(), is(title));

        List<String> tags = Arrays.asList(plainTags.split(" "));
        List<WebElement> tagElements = driver.findElements(By.cssSelector("doc > div.tags > ul > li"));
        for (WebElement each : tagElements) {
            String tag = each.findElement(By.cssSelector("a")).getText();
            assertThat(tags.contains(tag), is(true));
        }
    }

    public void answer(String answer) {
        driver.findElement(By.id("contents")).clear();
        driver.findElement(By.id("contents")).sendKeys(answer);
        driver.findElement(By.cssSelector(".btn-submit")).click();
    }

    public void verifyAnswer(String answer) {
        List<String> commentTexts = findCommentTexts();
        log.debug("comments size : {}", commentTexts.size());
//        assertThat(commentTexts.contains(answer), is(true));
    }

    private List<String> findCommentTexts() {
        List<WebElement> comments = driver.findElements(By.cssSelector("div.comment-doc"));
        List<String> commentTexts = Lists.newArrayList();
        for (WebElement comment : comments) {
            commentTexts.add(comment.findElement(By.cssSelector("p")).getText());
        }
        return commentTexts;
    }

    public void verifyAnswerCount(String answerCount) {
        String actual = driver.findElement(By.cssSelector(".qna-comment-count > strong")).getText();
        assertThat(actual, is(answerCount));
    }

    public QuestionPage likeAnswer() {
        driver.findElement(By.cssSelector("button.btn-like-article > strong.like-count")).click();
        driver.manage().timeouts().implicitlyWait(3, TimeUnit.SECONDS);
        return new QuestionPage(driver);
    }

    public void verifyLikeCount(int likeCount) {
        String actual = driver.findElement(By.cssSelector("button.btn-like-article > strong.like-count")).getText();
        assertThat(actual, is(likeCount + ""));
    }

    public AnswerUpdateFormPage goToUpdateAnswerPage() {
        driver.findElement(By.cssSelector("a.link-modify-article")).click();
        return new AnswerUpdateFormPage(driver);
    }

    public QuestionFormPage goToUpdatePage() {
        driver.findElement(By.cssSelector("a.link-modify-article")).click();
        return new QuestionFormPage(driver);
    }

    public void verifyBestAnswer() {
        String answerBest = driver.findElement(By.cssSelector(".qna-best-comment-label > strong")).getText();
        assertThat(answerBest, is("BEST 의견"));
    }

	public QuestionsPage goToQuestionsPage() {
		driver.findElement(By.cssSelector("nav.site-nav > ul > li")).click();
		return new QuestionsPage(driver);
	}
}

File: src/acceptance-test/scala/net/slipp/qna/QuestionPage.scala
package net.slipp.qna

import java.util.List
import java.util.concurrent.TimeUnit

import com.google.common.collect.Lists
import org.hamcrest.CoreMatchers.is
import org.junit.Assert._
import org.openqa.selenium.{By, WebDriver, WebElement}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.JavaConversions._

class QuestionPage(driver: WebDriver, title: String) {
  private val log: Logger = LoggerFactory.getLogger(classOf[QuestionPage])

  // assertThat(driver.getTitle, is(title))

  def this(driver: WebDriver) = this(driver, driver.getTitle)

  def verify(title: String, contents: String, plainTags: String) {
    val titleElement: WebElement = driver.findElement(By.cssSelector(".qna-title"))
    assertThat(titleElement.getText, is(title))
    val contentsElement = driver.findElement(By.cssSelector(".article-doc > p"))
    assertThat(contentsElement.getText, is(contents))
    verifyTags(plainTags)
  }

  def verifyTags(plainTags: String) = {
    val tags = plainTags.split(" ")
    val tagElements = driver.findElements(By.cssSelector("section.qna-tags > ul > li"))
    val addedTags = tagElements.map(e => e.findElement(By.cssSelector("a")).getText)
    tags.foreach(tag => {
      log.debug(s"tag : ${tag}")
      assertTrue(addedTags.contains(tag))
    })
  }

  def answer(answer: String) {
    driver.findElement(By.id("contents")).clear
    driver.findElement(By.id("contents")).sendKeys(answer)
    driver.findElement(By.cssSelector(".btn-submit")).click()
  }

  def verifyAnswer(answer: String) {
    val commentTexts: List[String] = findAnswerTexts
    log.debug("comments size : {}", commentTexts.size)
  }

  private def findAnswerTexts = {
    val comments = driver.findElements(By.cssSelector("div.comment-doc"))
    val commentTexts = Lists.newArrayList[String]()
    for (comment <- comments) {
      commentTexts.add(comment.findElement(By.cssSelector("p")).getText)
    }
    commentTexts
  }

  def deleteAnswer() {
    driver.findElement(By.cssSelector("button.link-delete-article")).click()
    
    val alert = driver.switchTo.alert
    alert.accept()
  }

  def verifyAnswerCount(answerCount: String) {
    val actual: String = driver.findElement(By.cssSelector(".qna-comment-count > strong")).getText
    assertThat(actual, is(answerCount))
    assertThat(findAnswerTexts.size(), is(answerCount.toInt))
  }

  def likeAnswer: QuestionPage = {
    driver.findElement(By.cssSelector("a.btn-like-article")).click()
    driver.manage.timeouts.implicitlyWait(3, TimeUnit.SECONDS)
    return new QuestionPage(driver)
  }

  def verifyLikeCount(likeCount: Int) {
    Thread.sleep(100)
    val actual: String = driver.findElement(By.cssSelector("a.btn-like-article > strong.like-article-count")).getText
    assertThat(actual, is(likeCount + ""))
  }

  def goToUpdateAnswerPage: AnswerUpdateFormPage = {
    driver.findElement(By.cssSelector("a.link-modify-article")).click()
    return new AnswerUpdateFormPage(driver)
  }

  def goToUpdatePage: QuestionFormPage = {
    driver.findElement(By.cssSelector("a.link-modify-article")).click()
    return new QuestionFormPage(driver)
  }

  def verifyBestAnswer {
    val answerBest: String = driver.findElement(By.cssSelector(".qna-best-comment-label > strong")).getText
    assertThat(answerBest, is("BEST 의견"))
  }

  def goToQuestionsPage = {
    driver.findElement(By.cssSelector("nav.site-nav > ul > li")).click()
    new QuestionsPage(driver)
  }

  def addTag(tag: String) = {
    val taggedForm = driver.findElement(By.id("taggedForm"))
    taggedForm.findElement(By.cssSelector("input.inp_nickname")).clear
    taggedForm.findElement(By.cssSelector("input.inp_nickname")).sendKeys(tag)
    taggedForm.findElement(By.cssSelector(".signin-with-sns-submit-btn")).click
    new QuestionPage(driver, driver.getTitle)
  }
}
